File: src/lib/scoring.ts
import type { LayerKind } from "@/lib/data/layers";
import type {
  AccessibilityEvaluation,
  AccessibilityFactor,
  AccessibilityProfileType,
  ConfidenceLevel,
} from "@/types";

export interface EntranceEvidence {
  name: string;
  steps: number;
  hasRamp: boolean;
  widthCm: number | null;
}

export interface FeatureEvidence {
  kind: LayerKind;
  status: string;
  stepCount?: number | null;
  verification?: "VERIFIED" | "COMMUNITY_REPORTED" | "UNKNOWN";
}

export interface ScoreInput {
  profile: AccessibilityProfileType;
  entrances: EntranceEvidence[];
  features: FeatureEvidence[];
  freshness: string;
}

export interface PlaceScaleScores {
  visual: AccessibilityEvaluation;
  mobility: AccessibilityEvaluation;
}

const BASE_SCORE = 50;
const VISUAL_KINDS: LayerKind[] = [
  "guiding_block",
  "pedestrian_crossing",
  "audio_crossing_signal",
  "obstacle",
  "surface_hazard",
];
const MOBILITY_FEATURE_KINDS: LayerKind[] = [
  "ramp",
  "stairs",
  "elevator",
  "path_width",
  "surface_condition",
  "accessible_entrance",
];

export function scoreLabel(score: number | null): AccessibilityEvaluation["label"] {
  if (score == null) return "Data Belum Tersedia";
  if (score >= 80) return "Sangat Aksesibel";
  if (score >= 60) return "Aksesibel Sebagian";
  if (score >= 40) return "Aksesibilitas Terbatas";
  return "Hambatan Signifikan";
}

export function scoreLevel(score: number | null): AccessibilityEvaluation["level"] {
  if (score == null) return "unknown";
  if (score >= 80) return "accessible";
  if (score >= 40) return "limited";
  return "not-accessible";
}

function clamp(value: number): number {
  return Math.max(0, Math.min(100, Math.round(value)));
}

function confidenceFor(count: number, hasVerified: boolean): ConfidenceLevel | null {
  if (count === 0) return null;
  if (count >= 4 && hasVerified) return "high";
  if (count >= 2) return "medium";
  return "low";
}

function entranceDeltas(entrance: EntranceEvidence): { delta: number; factors: AccessibilityFactor[] } {
  if (entrance.steps === 0) {
    return {
      delta: 1,
      factors: [{ label: `${entrance.name}: tanpa tangga`, kind: "positive" }],
    };
  }
  if (entrance.hasRamp && entrance.steps <= 2) {
    return {
      delta: 0.5,
      factors: [{ label: `${entrance.name}: ${entrance.steps} anak tangga dengan ramp`, kind: "neutral" }],
    };
  }
  return {
    delta: -1,
    factors: [{ label: `${entrance.name}: ${entrance.steps} anak tangga tanpa ramp`, kind: "negative" }],
  };
}

function entranceWidthFactor(entrance: EntranceEvidence): AccessibilityFactor[] {
  if (entrance.widthCm == null) return [];
  if (entrance.widthCm >= 120) {
    return [{ label: `${entrance.name}: pintu aksesibel ${entrance.widthCm} cm`, kind: "positive" }];
  }
  return [{ label: `${entrance.name}: pintu sempit ${entrance.widthCm} cm`, kind: "negative" }];
}

interface FeatureRule {
  kinds: LayerKind[];
  adjust: (feature: FeatureEvidence) => { delta: number; factors: AccessibilityFactor[] };
}

function numberOrNull(value: number | null | undefined): number | null {
  return typeof value === "number" ? value : null;
}

const MOBILITY_RULES: FeatureRule[] = [
  {
    kinds: ["ramp"],
    adjust: (f) =>
      f.status === "available"
        ? { delta: 1.2, factors: [{ label: "Ramp tersedia di area sekitar", kind: "positive" }] }
        : { delta: -1.2, factors: [{ label: "Ramp rusak di area sekitar", kind: "negative" }] },
  },
  {
    kinds: ["stairs"],
    adjust: (f) => ({
      delta: -1.5,
      factors: [
        {
          label: numberOrNull(f.stepCount) != null ? `Tangga ${f.stepCount} anak tangga` : "Ada tangga di jalur",
          kind: "negative",
        },
      ],
    }),
  },
  {
    kinds: ["elevator"],
    adjust: (f) =>
      f.status === "available"
        ? { delta: 1.2, factors: [{ label: "Elevator tersedia", kind: "positive" }] }
        : f.status === "out_of_service"
          ? { delta: -1.2, factors: [{ label: "Elevator sedang diperbaiki", kind: "negative" }] }
          : { delta: -1, factors: [{ label: "Elevator tidak tersedia", kind: "negative" }] },
  },
  {
    kinds: ["path_width"],
    adjust: (f) =>
      f.status === "accessible"
        ? { delta: 1, factors: [{ label: "Lebar jalur aksesibel", kind: "positive" }] }
        : f.status === "limited"
          ? { delta: -0.8, factors: [{ label: "Lebar jalur terbatas", kind: "neutral" }] }
          : { delta: -1.2, factors: [{ label: "Jalur terlalu sempit", kind: "negative" }] },
  },
  {
    kinds: ["surface_condition"],
    adjust: (f) =>
      f.status === "good"
        ? { delta: 0.8, factors: [{ label: "Permukaan jalur baik", kind: "positive" }] }
        : f.status === "uneven"
          ? { delta: -0.8, factors: [{ label: "Permukaan tidak rata", kind: "neutral" }] }
          : { delta: -1, factors: [{ label: "Permukaan jalur rusak", kind: "negative" }] },
  },
  {
    kinds: ["accessible_entrance"],
    adjust: (f) =>
      f.status === "accessible"
        ? { delta: 1.2, factors: [{ label: "Akses masuk aksesibel", kind: "positive" }] }
        : f.status === "partially_accessible"
          ? { delta: 0, factors: [{ label: "Akses masuk sebagian aksesibel", kind: "neutral" }] }
          : { delta: -1.5, factors: [{ label: "Akses masuk tidak aksesibel", kind: "negative" }] },
  },
];

const VISUAL_RULES: FeatureRule[] = [
  {
    kinds: ["guiding_block"],
    adjust: (f) =>
      f.status === "available"
        ? { delta: 1.2, factors: [{ label: "Guiding block tersedia", kind: "positive" }] }
        : f.status === "damaged"
          ? { delta: -1, factors: [{ label: "Guiding block rusak", kind: "negative" }] }
          : { delta: -1.2, factors: [{ label: "Guiding block terputus", kind: "negative" }] },
  },
  {
    kinds: ["pedestrian_crossing"],
    adjust: (f) =>
      f.status === "signalized" || f.status === "available"
        ? { delta: 1, factors: [{ label: "Zebra crossing dengan lampu", kind: "positive" }] }
        : f.status === "non_signalized"
          ? { delta: -0.5, factors: [{ label: "Zebra crossing tanpa lampu", kind: "neutral" }] }
          : { delta: -0.8, factors: [{ label: "Zebra crossing tidak tersedia", kind: "negative" }] },
  },
  {
    kinds: ["audio_crossing_signal"],
    adjust: (f) =>
      f.status === "available"
        ? { delta: 1.2, factors: [{ label: "Sinyal suara penyeberangan tersedia", kind: "positive" }] }
        : { delta: -1, factors: [{ label: "Sinyal suara penyeberangan tidak tersedia", kind: "negative" }] },
  },
  {
    kinds: ["obstacle"],
    adjust: (f) => ({
      delta: -1,
      factors: [{ label: `Hambatan di jalur (${f.status})`, kind: "negative" }],
    }),
  },
  {
    kinds: ["surface_hazard"],
    adjust: (f) => ({
      delta: -1,
      factors: [{ label: `Bahaya permukaan: ${f.status}`, kind: "negative" }],
    }),
  },
];

export function evaluateAccessibility(input: ScoreInput): AccessibilityEvaluation {
  const { profile, entrances, features, freshness } = input;
  const rules = profile === "WHEELCHAIR_MOBILITY" ? MOBILITY_RULES : VISUAL_RULES;
  const featureKinds = profile === "WHEELCHAIR_MOBILITY" ? MOBILITY_FEATURE_KINDS : VISUAL_KINDS;

  let delta = 0;
  const factors: AccessibilityFactor[] = [];
  let verifiedCount = 0;

  if (profile === "WHEELCHAIR_MOBILITY") {
    for (const entrance of entrances) {
      const entranceResult = entranceDeltas(entrance);
      delta += entranceResult.delta;
      factors.push(...entranceResult.factors, ...entranceWidthFactor(entrance));
    }
  }

  const relevant = features.filter((f) => featureKinds.includes(f.kind));
  for (const feature of relevant) {
    for (const rule of rules) {
      if (!rule.kinds.includes(feature.kind)) continue;
      const result = rule.adjust(feature);
      delta += result.delta;
      factors.push(...result.factors);
      if (feature.verification === "VERIFIED") verifiedCount += 1;
      break;
    }
  }

  const score = relevant.length + (profile === "WHEELCHAIR_MOBILITY" ? entrances.length : 0) === 0 ? null : clamp(BASE_SCORE + delta);
  const confidence = confidenceFor(
    factors.length,
    verifiedCount > 0,
  );

  return {
    score,
    label: scoreLabel(score),
    level: scoreLevel(score),
    factors,
    confidence,
    freshness,
    scale: 100,
  };
}

File: backend/internal/service/scoring.go
package service

import (
	"math"

	"anapsi/backend/internal/model"
)

const baseScore = 50

var visualKinds = []string{"guiding_block", "pedestrian_crossing", "audio_crossing_signal", "obstacle", "surface_hazard"}
var mobilityKinds = []string{"ramp", "stairs", "elevator", "path_width", "surface_condition", "accessible_entrance"}

// EntranceEvidence describes one entrance for scoring.
type EntranceEvidence struct {
	Name    string
	Steps   int
	HasRamp bool
	WidthCm *int
}

// FeatureEvidence describes one map feature for scoring.
type FeatureEvidence struct {
	Kind         string
	Status       string
	StepCount    *int
	Verification string
}

// AccessFactor is a human-readable scoring factor.
type AccessFactor struct {
	Label string `json:"label"`
	Kind  string `json:"kind"`
}

// AccessEvaluation is the scoring result for one place/profile.
type AccessEvaluation struct {
	Score      *int           `json:"score"`
	Label      string         `json:"label"`
	Level      string         `json:"level"`
	Factors    []AccessFactor `json:"factors"`
	Confidence *string        `json:"confidence"`
	Freshness  string         `json:"freshness"`
	Scale      int            `json:"scale"`
}

func scoreLabel(score *int) string {
	if score == nil {
		return "Data Belum Tersedia"
	}
	if *score >= 80 {
		return "Sangat Aksesibel"
	}
	if *score >= 60 {
		return "Aksesibel Sebagian"
	}
	if *score >= 40 {
		return "Aksesibilitas Terbatas"
	}
	return "Hambatan Signifikan"
}

func scoreLevel(score *int) string {
	if score == nil {
		return "unknown"
	}
	if *score >= 80 {
		return "accessible"
	}
	if *score >= 40 {
		return "limited"
	}
	return "not-accessible"
}

func clampScore(value float64) int {
	return int(math.Max(0, math.Min(100, math.Round(value))))
}

func confidenceFor(count int, hasVerified bool) *string {
	if count == 0 {
		return nil
	}
	if count >= 4 && hasVerified {
		s := "high"
		return &s
	}
	if count >= 2 {
		s := "medium"
		return &s
	}
	s := "low"
	return &s
}

func entranceDeltas(entrance EntranceEvidence) (float64, []AccessFactor) {
	if entrance.Steps == 0 {
		return 1, []AccessFactor{{Label: entrance.Name + ": tanpa tangga", Kind: "positive"}}
	}
	if entrance.HasRamp && entrance.Steps <= 2 {
		return 0.5, []AccessFactor{{Label: entrance.Name + ": " + model.FormatInt(entrance.Steps) + " anak tangga dengan ramp", Kind: "neutral"}}
	}
	return -1, []AccessFactor{{Label: entrance.Name + ": " + model.FormatInt(entrance.Steps) + " anak tangga tanpa ramp", Kind: "negative"}}
}

func entranceWidthFactor(entrance EntranceEvidence) []AccessFactor {
	if entrance.WidthCm == nil {
		return nil
	}
	if *entrance.WidthCm >= 120 {
		return []AccessFactor{{Label: entrance.Name + ": pintu aksesibel " + model.FormatInt(*entrance.WidthCm) + " cm", Kind: "positive"}}
	}
	return []AccessFactor{{Label: entrance.Name + ": pintu sempit " + model.FormatInt(*entrance.WidthCm) + " cm", Kind: "negative"}}
}

type featureRule struct {
	kinds  []string
	adjust func(f FeatureEvidence) (float64, []AccessFactor)
}

func numberOrNull(v *int) *int {
	if v == nil {
		return nil
	}
	return v
}

var mobilityRules = []featureRule{
	{
		kinds: []string{"ramp"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			if f.Status == "available" {
				return 1.2, []AccessFactor{{Label: "Ramp tersedia di area sekitar", Kind: "positive"}}
			}
			return -1.2, []AccessFactor{{Label: "Ramp rusak di area sekitar", Kind: "negative"}}
		},
	},
	{
		kinds: []string{"stairs"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			label := "Ada tangga di jalur"
			if numberOrNull(f.StepCount) != nil {
				label = "Tangga " + model.FormatInt(*f.StepCount) + " anak tangga"
			}
			return -1.5, []AccessFactor{{Label: label, Kind: "negative"}}
		},
	},
	{
		kinds: []string{"elevator"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			switch f.Status {
			case "available":
				return 1.2, []AccessFactor{{Label: "Elevator tersedia", Kind: "positive"}}
			case "out_of_service":
				return -1.2, []AccessFactor{{Label: "Elevator sedang diperbaiki", Kind: "negative"}}
			default:
				return -1, []AccessFactor{{Label: "Elevator tidak tersedia", Kind: "negative"}}
			}
		},
	},
	{
		kinds: []string{"path_width"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			switch f.Status {
			case "accessible":
				return 1, []AccessFactor{{Label: "Lebar jalur aksesibel", Kind: "positive"}}
			case "limited":
				return -0.8, []AccessFactor{{Label: "Lebar jalur terbatas", Kind: "neutral"}}
			default:
				return -1.2, []AccessFactor{{Label: "Jalur terlalu sempit", Kind: "negative"}}
			}
		},
	},
	{
		kinds: []string{"surface_condition"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			switch f.Status {
			case "good":
				return 0.8, []AccessFactor{{Label: "Permukaan jalur baik", Kind: "positive"}}
			case "uneven":
				return -0.8, []AccessFactor{{Label: "Permukaan tidak rata", Kind: "neutral"}}
			default:
				return -1, []AccessFactor{{Label: "Permukaan jalur rusak", Kind: "negative"}}
			}
		},
	},
	{
		kinds: []string{"accessible_entrance"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			switch f.Status {
			case "accessible":
				return 1.2, []AccessFactor{{Label: "Akses masuk aksesibel", Kind: "positive"}}
			case "partially_accessible":
				return 0, []AccessFactor{{Label: "Akses masuk sebagian aksesibel", Kind: "neutral"}}
			default:
				return -1.5, []AccessFactor{{Label: "Akses masuk tidak aksesibel", Kind: "negative"}}
			}
		},
	},
}

var visualRules = []featureRule{
	{
		kinds: []string{"guiding_block"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			switch f.Status {
			case "available":
				return 1.2, []AccessFactor{{Label: "Guiding block tersedia", Kind: "positive"}}
			case "damaged":
				return -1, []AccessFactor{{Label: "Guiding block rusak", Kind: "negative"}}
			default:
				return -1.2, []AccessFactor{{Label: "Guiding block terputus", Kind: "negative"}}
			}
		},
	},
	{
		kinds: []string{"pedestrian_crossing"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			if f.Status == "signalized" || f.Status == "available" {
				return 1, []AccessFactor{{Label: "Zebra crossing dengan lampu", Kind: "positive"}}
			}
			if f.Status == "non_signalized" {
				return -0.5, []AccessFactor{{Label: "Zebra crossing tanpa lampu", Kind: "neutral"}}
			}
			return -0.8, []AccessFactor{{Label: "Zebra crossing tidak tersedia", Kind: "negative"}}
		},
	},
	{
		kinds: []string{"audio_crossing_signal"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			if f.Status == "available" {
				return 1.2, []AccessFactor{{Label: "Sinyal suara penyeberangan tersedia", Kind: "positive"}}
			}
			return -1, []AccessFactor{{Label: "Sinyal suara penyeberangan tidak tersedia", Kind: "negative"}}
		},
	},
	{
		kinds: []string{"obstacle"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			return -1, []AccessFactor{{Label: "Hambatan di jalur (" + f.Status + ")", Kind: "negative"}}
		},
	},
	{
		kinds: []string{"surface_hazard"},
		adjust: func(f FeatureEvidence) (float64, []AccessFactor) {
			return -1, []AccessFactor{{Label: "Bahaya permukaan: " + f.Status, Kind: "negative"}}
		},
	},
}

// EvaluateAccessibility scores a place for a profile.
func EvaluateAccessibility(profile string, entrances []EntranceEvidence, features []FeatureEvidence, freshness string) AccessEvaluation {
	rules := mobilityRules
	featureKinds := mobilityKinds
	if profile != "WHEELCHAIR_MOBILITY" {
		rules = visualRules
		featureKinds = visualKinds
	}

	delta := 0.0
	var factors []AccessFactor
	verifiedCount := 0

	if profile == "WHEELCHAIR_MOBILITY" {
		for _, entrance := range entrances {
			entranceResult, entranceFactors := entranceDeltas(entrance)
			delta += entranceResult
			factors = append(factors, entranceFactors...)
			factors = append(factors, entranceWidthFactor(entrance)...)
		}
	}

	var relevant []FeatureEvidence
	for _, f := range features {
		for _, kind := range featureKinds {
			if f.Kind == kind {
				relevant = append(relevant, f)
				break
			}
		}
	}

	for _, feature := range relevant {
		for _, rule := range rules {
			matched := false
			for _, kind := range rule.kinds {
				if feature.Kind == kind {
					matched = true
					break
				}
			}
			if !matched {
				continue
			}
			result, ruleFactors := rule.adjust(feature)
			delta += result
			factors = append(factors, ruleFactors...)
			if feature.Verification == "VERIFIED" {
				verifiedCount++
			}
			break
		}
	}

	evidenceCount := len(relevant)
	if profile == "WHEELCHAIR_MOBILITY" {
		evidenceCount += len(entrances)
	}
	var score *int
	if evidenceCount != 0 {
		s := clampScore(baseScore + delta)
		score = &s
	}
	confidence := confidenceFor(len(factors), verifiedCount > 0)

	return AccessEvaluation{
		Score:      score,
		Label:      scoreLabel(score),
		Level:      scoreLevel(score),
		Factors:    factors,
		Confidence: confidence,
		Freshness:  freshness,
		Scale:      100,
	}
}

// SummaryScore resolves the display score for a profile.
func SummaryScore(score map[string]*int, profile string) (int, string) {
	if profile == "WHEELCHAIR_MOBILITY" {
		value := score["mobility"]
		if value == nil {
			return 0, "Belum dinilai"
		}
		return *value, ScoreToLabel(*value)
	}
	value := score["visual"]
	if value == nil {
		return 0, "Belum dinilai"
	}
	return *value, ScoreToLabel(*value)
}

// ScoreToLabel returns the human score label.
func ScoreToLabel(score int) string {
	if score >= 80 {
		return "Sangat Aksesibel"
	}
	if score >= 60 {
		return "Aksesibel Sebagian"
	}
	if score >= 40 {
		return "Aksesibilitas Terbatas"
	}
	return "Hambatan Signifikan"
}
